libraries/tools/kotlin-gradle-plugin/src/common/kotlin/org/jetbrains/kotlin/gradle/plugin/mpp/CompilationDetails.kt

```diff
@@ -527,10 +528,29 @@ class AndroidCompilationDetails(
     override val friendArtifacts: FileCollection
         get() = target.project.files(super.friendArtifacts, compilation.testedVariantArtifacts)
 
+    /*
+    * Example of how multiplatform dependencies from common would get to Android test classpath:
+    * commonMainImplementation -> androidDebugImplementation -> debugImplementation -> debugAndroidTestCompileClasspath
+    * After the fix for KT-35916 MPP compilation configurations receive a 'compilation' postfix for disambiguation.
+    * androidDebugImplementation remains a source set configuration, but no longer contains compilation dependencies.
+    * Therefore, it doesn't get dependencies from common source sets.
+    * We now explicitly add associate compilation dependencies to the Kotlin test compilation configurations (test classpaths).
+    * This helps, because the Android test classpath configurations extend from the Kotlin test compilations' directly.
+    */
     override fun addAssociateCompilationDependencies(other: KotlinCompilation<*>) {
-        if ((other as? KotlinJvmAndroidCompilation)?.androidVariant != getTestedVariantData(androidVariant)) {
-            super.addAssociateCompilationDependencies(other)
-        } // otherwise, do nothing: the Android Gradle plugin adds these dependencies for us, we don't need to add them to the classpath
+        compilation.compileDependencyConfigurationName.addAllDependenciesFromOtherConfigurations(
+            project,
+            other.apiConfigurationName,
+            other.implementationConfigurationName,
+            other.compileOnlyConfigurationName
+        )
+
+        compilation.runtimeDependencyConfigurationName.addAllDependenciesFromOtherConfigurations(
```

Comment:
Runtime classpath for tests is a bit more tricky in AGP - https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-main:build-system/gradle-core/src/main/java/com/android/build/gradle/internal/dependency/VariantDependenciesBuilder.java;l=299?q=VariantDependenciesBuild.

For unit test variants and library project device tests, we just add a dependency on "project" which publishes to RUNTIME_ELEMENTS.

For application project device tests, we don't add a dependency on "project" as it leads to duplicate classes across tested and test APK (causing runtime issues).

My proposal would be (for both compile and runtime):
- avoid adding configurations directly coming from the tested compilation
- add only transitive configurations for the test Android compilation (in order to get dependencies from "commonMain"  in Android device tests & unit tests).

Reply:
Interestingly enough, it seems that it's not necessary adding those dependencies to the runtimeDependencyConfiguration. Do you know who might be responsible for adding those dependencies to unit and instrumented tests in this case?

Reply:
That's AGP. For library projects, we add `project` dependency which is part of RUNTIME_ELEMENTS for the main Android variant i.e it contains dependencies from `commonMain` IIUC.

I think you can fully remove handling of the runtime classpath then.

Reply:
Runtime classpath handling fully removed ✅